File: results/0.5/DNS/dns-gpt-4-turbo.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>

// Define DNS message structure constants
#define QR_MASK 0x8000
#define OPCODE_MASK 0x7800
#define AA_MASK 0x0400
#define TC_MASK 0x0200
#define RD_MASK 0x0100
#define RA_MASK 0x0080
#define Z_MASK 0x0070
#define RCODE_MASK 0x000F

// DNS types
#define TYPE_A 1
#define TYPE_NS 2
#define TYPE_CNAME 5
#define TYPE_SOA 6
#define TYPE_PTR 12
#define TYPE_MX 15
#define TYPE_TXT 16
#define TYPE_AAAA 28

// DNS classes
#define CLASS_IN 1

// Function prototypes
static HParser *dns_header();
static HParser *dns_question();
static HParser *dns_rr();
static HParser *dns_message();

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <dns_packet_file>\n", argv[0]);
        return 1;
    }

    FILE *file = fopen(argv[1], "rb");
    if (!file) {
        perror("Failed to open file");
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long fsize = ftell(file);
    fseek(file, 0, SEEK_SET);

    uint8_t *buf = malloc(fsize);
    if (!buf) {
        perror("Failed to allocate memory");
        fclose(file);
        return 1;
    }

    fread(buf, 1, fsize, file);
    fclose(file);

    HParser *parser = dns_message();
    HParseResult *result = h_parse(parser, buf, fsize);
    if (result) {
        printf("DNS message parsed successfully.\n");
    } else {
        fprintf(stderr, "Failed to parse DNS message.\n");
    }

    h_parse_result_free(result);
    free(buf);
    h_parser_unref(parser);

    return 0;
}

static HParser *dns_header() {
    return h_sequence(
        h_uint16(), // ID
        h_bits(1, false), // QR
        h_bits(4, false), // Opcode
        h_bits(1, false), // AA
        h_bits(1, false), // TC
        h_bits(1, false), // RD
        h_bits(1, false), // RA
        h_bits(3, false), // Z
        h_bits(4, false), // RCODE
        h_uint16(), // QDCOUNT
        h_uint16(), // ANCOUNT
        h_uint16(), // NSCOUNT
        h_uint16(), // ARCOUNT
        NULL);
}

static HParser *dns_question() {
    return h_sequence(
        h_indirect(), // QNAME
        h_uint16(),   // QTYPE
        h_uint16(),   // QCLASS
        NULL);
}

static HParser *dns_rr() {
    return h_sequence(
        h_indirect(), // NAME
        h_uint16(),   // TYPE
        h_uint16(),   // CLASS
        h_uint32(),   // TTL
        h_length_value(h_uint16(), h_bits(0, false)), // RDLENGTH and RDATA
        NULL);
}

static HParser *dns_message() {
    return h_sequence(
        dns_header(),
        h_many(dns_question()), // Questions
        h_many(dns_rr()),       // Answer RRs
        h_many(dns_rr()),       // Authority RRs
        h_many(dns_rr()),       // Additional RRs
        NULL);
}

File: results/0.5/DNS/dns-gpt-4-turbo.rs
use nom::{
    bytes::complete::{take},
    multi::{count},
    number::complete::{be_u16, be_u32, be_u8},
    sequence::tuple,
    IResult,
};
use std::env;
use std::fs::File;
use std::io::{self, Read};

#[derive(Debug)]
struct Header {
    id: u16,
    qr: u8,
    opcode: u8,
    aa: u8,
    tc: u8,
    rd: u8,
    ra: u8,
    z: u8,
    rcode: u8,
    qdcount: u16,
    ancount: u16,
    nscount: u16,
    arcount: u16,
}

#[derive(Debug)]
struct Question {
    qname: Vec<u8>,
    qtype: u16,
    qclass: u16,
}

#[derive(Debug)]
struct ResourceRecord {
    name: Vec<u8>,
    rtype: u16,
    class: u16,
    ttl: u32,
    rdlength: u16,
    rdata: Vec<u8>,
}

#[derive(Debug)]
struct DnsPacket {
    header: Header,
    questions: Vec<Question>,
    answers: Vec<ResourceRecord>,
    authority: Vec<ResourceRecord>,
    additional: Vec<ResourceRecord>,
}

fn parse_name(input: &[u8]) -> IResult<&[u8], Vec<u8>> {
    let (input, length) = be_u8(input)?;
    if length == 0 {
        return Ok((input, vec![]));
    }
    let (input, name) = take(length)(input)?;
    let (input, mut subname) = parse_name(input)?;
    let mut full_name = vec![length];
    full_name.extend_from_slice(name);
    full_name.append(&mut subname);
    Ok((input, full_name))
}

fn parse_header(input: &[u8]) -> IResult<&[u8], Header> {
    let (input, (id, flags, qdcount, ancount, nscount, arcount)) = tuple((be_u16, be_u16, be_u16, be_u16, be_u16, be_u16))(input)?;
    let qr = (flags >> 15) & 0x1;
    let opcode = (flags >> 11) & 0xF;
    let aa = (flags >> 10) & 0x1;
    let tc = (flags >> 9) & 0x1;
    let rd = (flags >> 8) & 0x1;
    let ra = (flags >> 7) & 0x1;
    let z = (flags >> 4) & 0x7;
    let rcode = flags & 0xF;
    Ok((
        input,
        Header {
            id,
            qr: qr as u8,
            opcode: opcode as u8,
            aa: aa as u8,
            tc: tc as u8,
            rd: rd as u8,
            ra: ra as u8,
            z: z as u8,
            rcode: rcode as u8,
            qdcount,
            ancount,
            nscount,
            arcount,
        },
    ))
}

fn parse_question(input: &[u8]) -> IResult<&[u8], Question> {
    let (input, qname) = parse_name(input)?;
    let (input, (qtype, qclass)) = tuple((be_u16, be_u16))(input)?;
    Ok((input, Question { qname, qtype, qclass }))
}

fn parse_rr(input: &[u8]) -> IResult<&[u8], ResourceRecord> {
    let (input, name) = parse_name(input)?;
    let (input, (rtype, class, ttl, rdlength)) = tuple((be_u16, be_u16, be_u32, be_u16))(input)?;
    let (input, rdata) = take(rdlength)(input)?;
    Ok((
        input,
        ResourceRecord {
            name,
            rtype,
            class,
            ttl,
            rdlength,
            rdata: rdata.to_vec(),
        },
    ))
}

fn parse_dns_packet(input: &[u8]) -> IResult<&[u8], DnsPacket> {
    let (input, header) = parse_header(input)?;
    let (input, questions) = count(parse_question, header.qdcount as usize)(input)?;
    let (input, answers) = count(parse_rr, header.ancount as usize)(input)?;
    let (input, authority) = count(parse_rr, header.nscount as usize)(input)?;
    let (input, additional) = count(parse_rr, header.arcount as usize)(input)?;
    Ok((
        input,
        DnsPacket {
            header,
            questions,
            answers,
            authority,
            additional,
        },
    ))
}

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        return Err(io::Error::new(io::ErrorKind::InvalidInput, "No input file specified"));
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_dns_packet(&buffer) {
        Ok((_, packet)) => {
            println!("{:#?}", packet);
        }
        Err(e) => {
            println!("Failed to parse DNS packet: {:?}", e);
        }
    }

    Ok(())
}